### PR TITLE
Zotero 7-only plugin, submitted for review

### PR DESCRIPTION
**Pull request by aborel**
This PR is in response to issue https://github.com/UB-Mannheim/zotero-ocr/issues/52
Functionality is essentially the same as for the master branch, but supporting Zotero 7 instead of Zotero 6 and older.
I included a minor change that might also address https://github.com/UB-Mannheim/zotero-ocr/issues/46

### CODE REVIEW

**Comment by aborel** [comment]
My bad, the PR was a but hasty - I have messed up something while preparing it. I'll report again when it's fixed.

**Comment by aborel** [comment]
False alarm, the submitted code is actually OK (as far as I can tell). Feedback is not only welcome, but needed 😅

**Review by stweil — CHANGES_REQUESTED** [review]
I suggest to keep chrome, chrome.manifest and install.rdf (revert commit ee11b268b7ae651924d405ec5374cb46c6bb978b), but move them to the src subdirectory.

In addition either src/chrome/skin/default/zoteroocr/ubmannheim*png are required or the path for the icons has to be updated to avoid the duplicate png files. By the way, why do we have images 32 x 32 and 64 x 64, not 48 x 48 and 96 x 96 as in other add-ons?

Then the new plugin should work with Zotero 6 and with Zotero 7 (and we don't have to maintain a separate branch).

File: src/manifest.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  "version": "0.6.1",
+  "description": "Zotero Plugin for OCR",
+  "author": "Philipp Zumstein",
+  "icons": {
+    "32": "ubmannheim32x32.png",
+    "64": "ubmannheim64x64.png"
+  },
+  "applications": {
+    "zotero": {
+      "id": "zotero-ocr@bib.uni-mannheim.de",
```
Comment (stweil) [inline src/manifest.json:13]: @zuphilip, should we switch to zotero-ocr@uni-mannheim.de?

**Comment by aborel** [comment]
Thanks for the review! I'll take a closer look in a few days.

File: src/manifest.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      "update_url": "https://raw.githubusercontent.com/UB-Mannheim/zotero-ocr/master/update.json",
```
Comment (stweil) [inline src/manifest.json:14]:
```suggestion
      "update_url": "https://raw.githubusercontent.com/UB-Mannheim/zotero-ocr/master/updates.json",
```

File: src/prefs.xhtml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            <html:input type="text" id="pref-zoteroocr-pdftoppmPath-value" flex="1" preference="extensions.zotero.zoteroocr.pdftoppmPath"/>
+            <hbox>
+                <label value="Choose a language/script you want to use for recognition (default is eng):"/>
+                <html:input type="text" id="pref-zoteroocr-language-value" preference="extensions.zotero.zoteroocr.language" width="100"/>
+            </hbox>
+        </groupbox>
+        <groupbox>
+            <caption label="Output options"/>
+            <checkbox preference="extensions.zotero.zoteroocr.outputNote" label="Save output as a note"/>
+            <checkbox id="checkbox-zoteroocr-output-pdf" preference="extensions.zotero.zoteroocr.outputPDF" label="Save output as a PDF with text layer" oncommand="Zotero.OCR.updatePDFOverwritePref()"/>
```
Comment (stweil) [inline src/prefs.xhtml:16]: In Zotero 6 the OCR was triggered from the context menu. Using a checkbox now might be sufficient for a first test, but is not the right solution for Zotero 7.

**Review by stweil — CHANGES_REQUESTED** [review]

File: src/zotero-ocr.js [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+            if (!win.ZoteroPane) continue;
+            this.removeFromWindow(win);
+        }
+    },
+    
+    toggleGreen(window, enabled) {
+        window.document.documentElement
+            .toggleAttribute('data-green-instead', enabled);
+    },
+
```
Comment (stweil) [inline src/zotero-ocr.js:94]: This code should be removed as it is not used by Zotero OCR.

Comment (aborel) [reply src/zotero-ocr.js:94]: Forgotten code from the make-it-red example plugin, of course it is useless here.

File: src/zotero-ocr.js [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+        Zotero.debug("ZoteroOCR: " + msg);
+    },
+    
+    addToWindow(window) {
+        let doc = window.document;
+        
+        // Add a stylesheet to the main Zotero pane
+        /*
+        let link1 = doc.createElement('link');
+        link1.id = 'make-it-red-stylesheet';
```
[the remark this answers is not in the document]
Comment (aborel) [reply src/zotero-ocr.js:33]: OK.

**Comment by mpr1255** [comment]
Really looking forward to giving this a spin!!

File: src/zotero-ocr.js [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        }
+    },
+    
+    async main() {
+        // Global properties are included automatically in Zotero 7
+        var host = new URL('https://foo.com/path').host;
+        this.log(`Host is ${host}`);
+        
+        // Retrieve a global pref
+        this.log(`Intensity is ${Zotero.Prefs.get('extensions.make-it-red.intensity', true)}`);
```
[the remark this answers is not in the document]
Comment (aborel) [reply src/zotero-ocr.js:317]: OK

File: src/bootstrap.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	Zotero.PreferencePanes.register({
+		pluginID: 'zotero-ocr@bib.uni-mannheim.de',
+		src: rootURI + 'prefs.xhtml',
+		//scripts: [rootURI + 'prefs.js']
+	});
+	
+	Services.scriptloader.loadSubScript(rootURI + 'zotero-ocr.js');
+	ZoteroOCR.init({ id, version, rootURI });
+	ZoteroOCR.addToAllWindows();
+	await ZoteroOCR.main();
```
Comment (stweil) [inline src/bootstrap.js:23]:
```suggestion
```

File: src/zotero-ocr.js [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+                }
+            }
+        }
+    },
+    
+    async main() {
+        // Global properties are included automatically in Zotero 7
+        var host = new URL('https://foo.com/path').host;
+        
+    },
```
Comment (stweil) [inline src/zotero-ocr.js:323]:
```suggestion
```

Comment (stweil) [reply src/zotero-ocr.js:323]: `main` is a relict from copy+paste which is not needed here.

**Review by stweil — CHANGES_REQUESTED** [review]

File: src/bootstrap.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+async function startup({ id, version, rootURI }) {
+	log("Starting Zotero OCR");
+	
+	Zotero.PreferencePanes.register({
+		pluginID: 'zotero-ocr@bib.uni-mannheim.de',
+		src: rootURI + 'prefs.xhtml',
+		//scripts: [rootURI + 'prefs.js']
+	});
+
```
Comment (stweil) [inline src/bootstrap.js:19]:
```suggestion

```

Comment (stweil) [reply src/bootstrap.js:19]: This line (and some more in other files) add trailing whitespace which should be removed.

**Review by stweil — CHANGES_REQUESTED** [review]

File: src/style.css [hunk 10]
```diff
@@ -0,0 +1,7 @@
+.row {
```
Comment (stweil) [inline src/style.css:1]: This file is a copy+paste relict which can be removed.

**Review by stweil — CHANGES_REQUESTED** [review]

**Comment by stweil** [comment]
Can you squash the commits which remove .DS_Store files (a3ad7cbdf8933c9dde931a628391c6e78ddd51d1, ee11b268b7ae651924d405ec5374cb46c6bb978b) with their predecessors to clean the history a bit?

File: chrome.manifest [hunk 11]
```diff
@@ -1,7 +0,0 @@
-content zoteroocr chrome/content/
```
Comment (stweil) [inline chrome.manifest:1]: Please don't remove this file. It is required for a plugin which supports both Zotero 6 and Zotero 7.

File: chrome/content/overlay.xul [hunk 12]
```diff
@@ -1,32 +0,0 @@
-<?xml version="1.0"?>
```
Comment (stweil) [inline chrome/content/overlay.xul:1]: Please don't remove this file. It is required for a plugin which supports both Zotero 6 and Zotero 7.

File: chrome/content/preferences.xul [hunk 13]
```diff
@@ -1,61 +0,0 @@
-<?xml version="1.0"?>
```
Comment (stweil) [inline chrome/content/preferences.xul:1]: Please don't remove this file. It is required for a plugin which supports both Zotero 6 and Zotero 7.

File: chrome/content/zoteroocr.js [hunk 14]
```diff
@@ -1,234 +0,0 @@
-// zoteroocr.js
```
Comment (stweil) [inline chrome/content/zoteroocr.js:1]: Please don't remove this file. It is required for a plugin which supports both Zotero 6 and Zotero 7.

File: chrome/locale/en-US/zoteroocr.dtd [hunk 15]
```diff
@@ -1,2 +0,0 @@
-<!ENTITY ocr.menu.label			"OCR selected PDF(s)">
```
Comment (stweil) [inline chrome/locale/en-US/zoteroocr.dtd:1]: Please don't remove this file. It is required for a plugin which supports both Zotero 6 and Zotero 7.

File: install.rdf [hunk 16]
```diff
@@ -1,34 +0,0 @@
-<?xml version="1.0"?>
```
Comment (stweil) [inline install.rdf:1]: Please don't remove this file. It is required for a plugin which supports both Zotero 6 and Zotero 7.

File: update.rdf [hunk 17]
```diff
@@ -1,30 +0,0 @@
-<?xml version="1.0" encoding="utf-8" ?>
```
Comment (stweil) [inline update.rdf:1]: Please don't remove this file. It is required for a smooth update from the current Zotero 6 plugin to a new one which supports both Zotero 6 and Zotero 7.

**Review by stweil — CHANGES_REQUESTED** [review]

File: src/zotero-ocr.js [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+          id: "zotero-ocr-item-menu",
+          label: "Zotero-OCR" //,
+          //class: "menuitem-iconic",
+          //image: "chrome://zotero-better-bibtex/content/skin/bibtex-menu.svg"
+        })).appendChild(elements.create("menupopup"));
+        menupopup.appendChild(elements.create("menuitem", {
+          label: "Test OCR",
+          oncommand: () => ZoteroOCR.recognize(window)
+        }));
+        */
```
Comment (stweil) [inline src/zotero-ocr.js:68]: Remove that test code?

File: src/bootstrap.js [hunk 19]
```diff
@@ -0,0 +1,42 @@
+var ZoteroOCR;
+
+function log(msg) {
+	Zotero.debug("Zotero OCR: " + msg);
+}
+
+function install() {
+	log("Installed Zotero OCR");
```
Comment (stweil) [inline src/bootstrap.js:8]:
```suggestion
	log("Installed");
```

File: src/bootstrap.js [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+async function startup({ id, version, rootURI }) {
+	log("Starting Zotero OCR");
```
Comment (stweil) [inline src/bootstrap.js:12]:
```suggestion
	log("Starting");
```

File: src/bootstrap.js [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+function onMainWindowLoad({ window }) {
+	ZoteroOCR.addToWindow(window);
+}
+
+function onMainWindowUnload({ window }) {
+	ZoteroOCR.removeFromWindow(window);
+}
+
+function shutdown() {
+	log("Shutting down Zotero OCR");
```
Comment (stweil) [inline src/bootstrap.js:35]:
```suggestion
	log("Shutting down");
```

File: src/bootstrap.js [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+	ZoteroOCR.removeFromAllWindows();
+	ZoteroOCR = undefined;
+}
+
+function uninstall() {
+	log("Uninstalled Zotero OCR");
```
Comment (stweil) [inline src/bootstrap.js:41]:
```suggestion
	log("Uninstalled");
```

**Review by stweil — CHANGES_REQUESTED** [review]
The `log` method already writes "Zotero OCR", and there is no need to write it twice.

File: src/zotero-ocr.js [hunk 23]
```diff
@@ -0,0 +1,324 @@
...
+            ZoteroOCR.recognize(window);
+        });
+        doc.getElementById('menu_ToolsPopup').appendChild(menuitem);
+        this.storeAddedElement(menuitem);
+
+
+        let popupmenuitem = doc.createXULElement('menuitem');
+        popupmenuitem.id = 'zotero-ocr-item-menu';
+        popupmenuitem.class = 'menuitem-iconic zotero-menuitem-ocr'
+        //popupmenuitem.setAttribute('type', 'checkbox');
```
Comment (stweil) [inline src/zotero-ocr.js:49]:
```suggestion
```

File: src/manifest.json (hunk 1 above)
Comment (stweil) [reply src/manifest.json:13]: No, we better keep the current id.

**Comment by stweil** [comment]
@aborel, would it be okay if I squash your commits with all suggested changes, add updates to support Zotero 6 and 7 simultaneously and then create a new pull request for that?

File: src/updates.json [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+          "applications": {
+            "zotero": {
+              "strict_min_version": "6.999"
+            }
+          }
+        }
+      ]
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (aborel) [reply src/updates.json:17]: OK

File: src/manifest.json (hunk 2 above)
Comment (aborel) [reply src/manifest.json:14]: OK

File: src/manifest.json [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+      "strict_min_version": "6.999",
+      "strict_max_version": "7.0.*"
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (aborel) [reply src/manifest.json:19]: OK

File: src/zotero-ocr.js [hunk 26]
```diff
@@ -0,0 +1,319 @@
...
+Components.utils.import("resource://gre/modules/osfile.jsm");
+
+
+ZoteroOCR = {
+    id: null,
+    version: null,
+    rootURI: null,
+    initialized: false,
+    addedElementIDs: [],
+
```
[the remark this answers is not in the document]
Comment (aborel) [reply src/zotero-ocr.js:14]: OK

File: src/bootstrap.js (hunk 9 above)
Comment (aborel) [reply src/bootstrap.js]: OK

File: src/zotero-ocr.js (hunk 8 above)
Comment (aborel) [reply src/zotero-ocr.js]: I had assumed the function was necessary (although perhaps the content might not do anything useful), but if it isn't, of course let's get rid of it.

File: src/bootstrap.js (hunk 7 above)
Comment (aborel) [reply src/bootstrap.js:23]: OK

File: src/style.css (hunk 10 above)
Comment (aborel) [reply src/style.css]: OK

File: src/zotero-ocr.js [hunk 27]
```diff
@@ -0,0 +1,324 @@
...
+        menuitem.id = 'ocr-selected-pdfs';
+        menuitem.setAttribute('type', 'checkbox');
+        menuitem.setAttribute('data-l10n-id', 'ocr-selected-pdfs');
+        // MozMenuItem#checked is available in Zotero 7
+        menuitem.addEventListener('command', () => {
+            ZoteroOCR.recognize(window);
+        });
+        doc.getElementById('menu_ToolsPopup').appendChild(menuitem);
+        this.storeAddedElement(menuitem);
+
```
[the remark this answers is not in the document]
Comment (aborel) [reply src/zotero-ocr.js]: Yes, as mentioned at the time, I was leaving the code in an intermediate state. Done now.

File: src/zotero-ocr.js (hunk 18 above)
Comment (aborel) [reply src/zotero-ocr.js]: OK

File: chrome.manifest (hunk 11 above)
Comment (aborel) [reply chrome.manifest:1]: As discussed in the issue https://github.com/UB-Mannheim/zotero-ocr/issues/52#issuecomment-1975652990 , I am not interested in supporting Zotero 6 and I prefer to submit a Z7-only version on this side branch.

File: src/bootstrap.js (hunk 19 above)
Comment (aborel) [reply src/bootstrap.js:8]: OK

File: src/bootstrap.js (hunk 21 above)
Comment (aborel) [reply src/bootstrap.js:34]: OK

File: src/bootstrap.js (hunk 22 above)
Comment (aborel) [reply src/bootstrap.js:40]: OK

File: src/zotero-ocr.js (hunk 23 above)
Comment (aborel) [reply src/zotero-ocr.js]: OK

**Comment by aborel** [comment]
> @aborel, would it be okay if I squash your commits with all suggested changes (perhaps with a subject line "Add support for Zotero 7"), add updates to support Zotero 6 and 7 simultaneously and then create a new pull request for that?

Yes, no problem!

**Comment by stweil** [comment]
> > @aborel, would it be okay if I squash your commits with all suggested changes (perhaps with a subject line "Add support for Zotero 7"), add updates to support Zotero 6 and 7 simultaneously and then create a new pull request for that?
> 
> Yes, no problem!

See pull request #63 for the result.
It now also builds an xpi file which can be used for testing automatically (see [actions](https://github.com/UB-Mannheim/zotero-ocr/actions)).

**Comment by stweil** [comment]
Meanwhile the changes from this pull request were committed to the master branch in commit 2f8e4ca9e7499de6d17ab22764c37f7bb4959102, so they are part of the new [release 0.7.0](https://github.com/UB-Mannheim/zotero-ocr/releases/tag/0.7.0) which now supports Zotero 7, too.

Thanks a lot for this great contribution!

**Comment by zuphilip** [comment]
Superseded by #63

**Pull request closed without merging**